### ISSUE TITLE
Google Pixel 3a won't shut down propertly

### ISSUE THREAD

**Issue by cybercrypt13** [ubports/ubuntu-touch#1732]
Please see linked video for detailed description of issue and it happens pretty much 100% of the time.

**https://youtu.be/dDjBVlustek**

**Comment by fredldotme** [comment]
Thanks for the bug report. This is a known issue, but I'll take a look at it in the coming week. I will update this issue when there are fixes in place.

**Comment by Xozk** [comment]
I have to Power Off my Pixel 3a several times, everytime when I want it to turn off. Any fix for this?

**Comment by angelomaragna** [comment]
Google Pixel 3a XL here. Not sure if it's a software issue or a platform issue or a boot one.

Having seen @cybercrypt13 video, I realized mine was continuously restarting because I had the power cable connected. 
With power connected, the device never shuts down. No shutdown command seems to power it off and have it stay so. With UT started, holding down the power button and choosing "Power off", it gets its into a reboot cycle.

If I hold down "power" + "lower volume" keys and get into the boot menu, and then select "stop" the device restarts anyway. 
It seem to be impossible to turn it off. 

On the other side, if I disconnect the charger and click power off, after 2-3 times I do so, it effectively shuts down.


Happy to run any diagnostic or debug command if you have any.

**Comment by fredldotme** [comment]
I took a few hours of my spare time to look into that issue, and it seems that this particular behavior can't be easily solved on a 16.04 base with Upstart as pid 1.

Upstart services are in a state where they want to shutdown but can't because some event might not have fired, being unable to tear down the services successfully.

On 20.04 we are going to use systemd, which means proper teardown of service units, whether it being lxc-android-config (the Halium container) or UT-side services interfacing with it.

If someone is willing to put up with debugging it, then by all means go ahead, but I don't see a particular need to focus on that myself any more than I have already.

TL;DR:
This issue should be fixed or easier to solve using systemd on 20.04.

**Comment by fredldotme** [comment]
Ok, I just checked again and noticed something during shutdown. Implemented a fix here: https://github.com/fredldotme/android_kernel_google_bonito/commit/fbde55254d16faf260f6e252c2d9b6c00f525bc9

Should be available in the next day or two for testing.

**Comment by Flohack74** [comment]
@fredldotme can this be relevant for other devices as well?

**Comment by fredldotme** [comment]
Potentially, at least on QCOM devices (assuming pstore shows adsprpc hanging).
I had to revert the mentioned two changes and remove the hanging function call altogether to get call audio back:
https://github.com/fredldotme/android_kernel_google_bonito/commit/9ddf7e819770c419d57de12422fc5fe57dd070d6 

To find out whether your device is affected:

```
phablet@ubuntu-phablet:~$ cat /etc/sysctl.d/98-printk.conf 
kernel.printk = 8 8 8 8
```

Create the file with those contents, reboot and try to trigger the bug. pstore should give valuable information about what went wrong now.

**Comment by mateosalta** [comment]
tested pixel 3a dev 594. Turns off completely without charger. With charger - does a full power off before going into offline charging for a moment and then booting back up.

I set off-mode-charge to 1 and the same happened - is the device being told to reboot in charge mode? https://android.stackexchange.com/questions/20021/automatically-power-on-android-when-the-charger-is-connected
Idealy it should turn off while plugged in - then stay in charge mode until unplugged or manually booted.

Tldr: unplugged - fully fixed. plugged in - much better, does a full shutdown, but does end up booting the system again

**Comment by nitanmarcel** [comment]
> Potentially, at least on QCOM devices (assuming pstore shows adsprpc hanging).
> I had to revert the mentioned two changes and remove the hanging function call altogether to get call audio back:
> [fredldotme/android_kernel_google_bonito@9ddf7e8](https://github.com/fredldotme/android_kernel_google_bonito/commit/9ddf7e819770c419d57de12422fc5fe57dd070d6)
> 
> To find out whether your device is affected:
> 
> ```
> phablet@ubuntu-phablet:~$ cat /etc/sysctl.d/98-printk.conf 
> kernel.printk = 8 8 8 8
> ```
> 
> Create the file with those contents, reboot and try to trigger the bug. pstore should give valuable information about what went wrong now.

Sorry for the reply but I haven't understood something:

I need only this commit to fix the shutdown hanging or also the other 2? From the commit message it states that it fixes the hanging breaking the call audio (which I haven't noticed this, just he broken shutdown/restarts as you have mentioned in the issue details)

Thanks!

**Comment by fredldotme** [comment]
> > Potentially, at least on QCOM devices (assuming pstore shows adsprpc hanging).
> > I had to revert the mentioned two changes and remove the hanging function call altogether to get call audio back:
> > [fredldotme/android_kernel_google_bonito@9ddf7e8](https://github.com/fredldotme/android_kernel_google_bonito/commit/9ddf7e819770c419d57de12422fc5fe57dd070d6)
> > To find out whether your device is affected:
> > ```
> > phablet@ubuntu-phablet:~$ cat /etc/sysctl.d/98-printk.conf 
> > kernel.printk = 8 8 8 8
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Create the file with those contents, reboot and try to trigger the bug. pstore should give valuable information about what went wrong now.
> 
> Sorry for the reply but I haven't understood something:
> 
> I need only this commit to fix the shutdown hanging or also the other 2? From the commit message it states that it fixes the hanging breaking the call audio (which I haven't noticed this, just he broken shutdown/restarts as you have mentioned in the issue details)
> 
> Thanks!

You might only need this one here, as the "reverting the 2 commits" restored state back to normal: https://github.com/fredldotme/android_kernel_google_bonito/commit/9ddf7e819770c419d57de12422fc5fe57dd070d6
